Add topography to coarse mesh box

anne-glerum opened the pull request:
This adaptation adds topography to the box geometry coarse mesh. I have two issues I'd like some input on. First, at the time the coarse mesh is created the `time` value in simulator is still set to NaN, so I had to adapt the check in AsciiDataBoundary::get_data_component. Is this test for NaN ok to do? Second,  the topography is only added to the coarse mesh, during initial refinement it is not updated. Which means topography might not even be seen by the coarse mesh if no repetitions are specified. For geometries with a manifold this is not a problem, because the push forward/pull back functions include the topography and new points are constructed with these functions. Should we add the topography during initial refinement as well somehow for this box geometry?

gassmoeller commented:
What about adding another signal called `pre_set_initial_state` that is called around `core.cc:2363` after setting the timestep, but before setting the initial condition. Box would then need to have a function that is connected to this signal and calls `GridTools::transform`, and it would be called before every initial adaptive refinement (and only once if there is no adaptive initial refinement). This would also remove the necessity of testing the timestep against NaN, because it is always called *after* the timestep is already set. Would that achieve what you intended?

anne-glerum commented:
Hi @gassmoeller, I took your suggestion and added a signal. Could you maybe have a look?

gassmoeller commented:
Yes that is what I meant. If that works for you lets do it this way. Let me know when I should give this PR a proper review.

gassmoeller commented on include/aspect/geometry_model/box.h:
```diff
@@ -30,17 +30,39 @@ namespace aspect
 {
   namespace GeometryModel
   {
+
```
Please remove this empty line.

gassmoeller commented on include/aspect/geometry_model/box.h:
```diff
@@ -30,17 +30,39 @@ namespace aspect
+        /**
+         * Relocate the vertical coordinate of the given point based on
+         * the topography at the surface specified by the initial topography
+         * model.
```
Please describe the difference between the two functions below or separate the documentation into two blocks, one per function.

gassmoeller commented on include/aspect/geometry_model/box.h:
```diff
@@ -161,6 +183,7 @@ namespace aspect
         void
         parse_parameters (ParameterHandler &prm);
 
+
```
Another spare empty line

gassmoeller commented on source/geometry_model/box.cc:
```diff
@@ -21,6 +21,8 @@
 
 #include <aspect/geometry_model/box.h>
 #include <aspect/geometry_model/initial_topography_model/zero_topography.h>
+#include <aspect/geometry_model/initial_topography_model/interface.h>
```
I do not think you need this include, because you include `zero_topography` anyway.

gassmoeller commented on source/geometry_model/box.cc:
```diff
@@ -299,10 +349,22 @@ namespace aspect
         prm.leave_subsection();
       }
       prm.leave_subsection();
+
```
Please remove this empty line (it is unusual in other plugins).

gassmoeller commented on source/geometry_model/box.cc:
```diff
@@ -318,6 +380,15 @@ namespace aspect
   }
 }
+
+//ASPECT_REGISTER_SIGNALS_CONNECTOR(signal_connector<2>,signal_connector<3>)
```
These lines are not longer necessary

gassmoeller commented on source/geometry_model/box.cc:
```diff
@@ -318,6 +380,15 @@ namespace aspect
+                                   "It is possible to add a topography to the box model. Note however that "
+                                   "the representation of this topography depends greatly on the number "
+                                   "of repetitions specified for the coarse mesh, as the topography is not "
+                                   "updated in the initial global and adaptive refinement cycles.")
```
Maybe also add a sentence that the topography is supposed to be small (because it is not taken into account to compute depth or representative point).

anne-glerum commented:
Hi Rene, no problem, I could have just asked again :) I've made the changes you suggested, and some more, could you please check them? Thanks!

tjhei commented:
/run-tests

tjhei commented on source/geometry_model/box.cc:
```diff
@@ -300,9 +353,20 @@ namespace aspect
+
+    template <int dim>
+    void
+    Box<dim>::connect_to_signal (SimulatorSignals<dim> &signals)
```
this doesn't need to be a public function right? You are also only calling it once and it only contains a single statement. Maybe move it into ``initialize()``?

tjhei commented on source/geometry_model/box.cc:
```diff
@@ -318,6 +382,12 @@ namespace aspect
                                    "and top boundaries (see also the documentation of the deal.II class "
                                    "``GeometryInfo''). You can also use symbolic names ``left'', ``right'', "
-                                   "etc., to refer to these boundaries in input files.")
+                                   "etc., to refer to these boundaries in input files."
```
missing space at the end

gassmoeller commented:
Hi Anne, this seems essentially ready except for the two stylistic issues raised by Timo, and the conflict. I am currently cleaning things that lay around for far too long, and this is definitely one of them. Let me know, when this is ready.

anne-glerum commented:
Hi @gassmoeller , I think I addressed all comments. I didn't know whether to put the entry in changes.h at the original date or today, but went with the former. Hope it can be wrapped up now!

gassmoeller commented:
:+1:
